crates/subspace-farmer/src/cluster/nats_client.rs

```diff
@@ -849,3 +852,105 @@ fn subject_with_instance(subject: &'static str, instance: Option<&str>) -> Subje
         Subject::from_static(subject)
     }
 }
+
+/// Spawn a queue responder task that will process requests from the given subject
+/// using the provided processing function.
+///
+/// This will create a subscription on the subject for the given instance (if provided)
+/// and queue group. Incoming messages will be deserialized as the request type `GR`
+/// and passed to the `process` function to produce a response of type `GR::Response`.
+/// The response will then be sent back on the reply subject from the original request.
+///
+/// # Arguments
+///
+/// * `label` - A label to use for logging
+/// * `nats_client` - The NATS client to use for the subscription and publishing responses
+/// * `instance` - Optional instance name to use in place of the `*` in the subject
+/// * `group` - The queue group name for the subscription
+/// * `process` - The function to call with the decoded request to produce a response
+pub async fn generic_queue_responder<GR, F, OP>(
+    label: &'static str,
+    nats_client: &NatsClient,
+    instance: Option<&str>,
+    group: String,
+    process: OP,
+) -> anyhow::Result<()>
+where
+    GR: GenericRequest,
+    F: Future<Output = Option<GR::Response>> + Send,
+    OP: Fn(GR) -> F + Send + Sync,
+{
+    // Initialize with pending future so it never ends
+    let mut processing = FuturesUnordered::<Pin<Box<dyn Future<Output = ()> + Send>>>::from_iter([
+        Box::pin(pending()) as Pin<Box<_>>,
+    ]);
+
+    let subject = match instance {
+        Some(instance) => GR::SUBJECT.replace('*', instance),
+        None => GR::SUBJECT.to_string(),
+    };
+
+    let subscription = nats_client
+        .queue_subscribe(subject, group)
+        .await
+        .map_err(|error| {
+            anyhow!("Failed to subscribe to {label} requests for  {instance:?}: {error}",)
+        })?;
+    debug!(%label, "requests subscription");
+    let mut subscription = subscription.fuse();
+
+    loop {
+        select! {
+            maybe_message = subscription.next() => {
+                let Some(message) = maybe_message else {
+                    break;
+                };
+
+                // Create background task for concurrent processing
+                processing.push(Box::pin(process_generic_queue_request(
+                    label,
+                    nats_client,
+                    message,
+                    &process,
+                )));
+            }
+            _ = processing.next() => {
+                // Nothing to do here
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn process_generic_queue_request<GR, F, OP>(
+    label: &'static str,
+    nats_client: &NatsClient,
+    Message { reply, payload, .. }: Message,
```

Comment:
Tiny nit: I probably wouldn't de-structure it here, it makes function definition more cluttered in presence of generics and other arguments.

Reply:
fixed